[WIP] Use weak linkage instead of compiler generated shims

bjorn3 opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/134522)*
<!-- homu-ignore:end -->

https://github.com/rust-lang/rust/pull/86844 (unstably) made it possible to avoid the allocator shim when using `#[global_allocator]`. This PR makes it possible to also avoid the allocator shim when using the default allocator in libstd by making use of weak linkage. Eventual stabilization of avoiding the allocator shim may be blocked on it working with the default allocator too.

This is still keeping the `__rust_no_alloc_shim_is_unstable` symbol for now until https://github.com/rust-lang/rust/issues/123015 gets stabilized.

TODO: Update comments everywhere, test on <strike>macOS and</strike> Windows and write a better PR description why we want this. Also prevent codegen of weak symbols when there is a non-weak definition of the same symbol in the same codegen unit.

try-job: x86_64-msvc
try-zzz-job: x86_64-apple-1
try-zzz-job: x86_64-apple-2

bjorn3 committed:
[WIP] Use weak linkage instead of compiler generated shims

This is still keeping the allocator shim for the oom handler and the
__rust_no_alloc_shim_is_unstable symbol for now.

TODO: Update comments everywhere and test on macOS and Windows

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12416458111/job/34665169647) [(plain)](https://github.com/rust-lang/rust/commit/f4f69369028c12f459d91e5ff47c017fe2dd6397/checks/34665169647/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12416937163/job/34666746042) [(plain)](https://github.com/rust-lang/rust/commit/93bd5d3e3ba003da8011743a85654dba4df6d32a/checks/34666746042/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
....................................................................i....

failures:

---- library/core/src/alloc/global.rs - alloc::global::GlobalAlloc (line 22) stdout ----
error: the `#[global_allocator]` in this crate conflicts with global allocator in: std
error: aborting due to 1 previous error

Couldn't compile the test.

```

</details>

bjorn3 committed:
Update comments and simplify miri allocator handling a bit

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12428932385/job/34701463602) [(plain)](https://github.com/rust-lang/rust/commit/1b00966c2579c9db89800826761cf3bb20d47fb3/checks/34701463602/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
....................................................................i....

failures:

---- library/core/src/alloc/global.rs - alloc::global::GlobalAlloc (line 22) stdout ----
error: the `#[global_allocator]` in this crate conflicts with global allocator in: std
error: aborting due to 1 previous error

Couldn't compile the test.

```

</details>

bjorn3 committed:
Use weak linkage for the alloc error handler too

bjorn3 committed:
Revert some miri changes

bjorn3 committed:
Provide a default for __rust_alloc_error_handler_should_panic

bjorn3 commented:
@bors try

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/12430755919/job/34706813549) [(plain)](https://github.com/rust-lang-ci/rust/commit/25f4036c4dfd6feb59ebd8be861c909aaf2973e2/checks/34706813549/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Updating files:  99% (50521/50914)
Updating files:  99% (50812/50914)
Updating files: 100% (50914/50914)
Updating files: 100% (50914/50914), done.
branch 'try' set up to track 'origin/try'.
Switched to a new branch 'try'
[command]"C:\Program Files\Git\bin\git.exe" log -1 --format=%H
25f4036c4dfd6feb59ebd8be861c909aaf2973e2
##[group]Run src/ci/scripts/setup-environment.sh
src/ci/scripts/setup-environment.sh
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12430626379/job/34706443170) [(plain)](https://github.com/rust-lang/rust/commit/abd406c2ca123f21598409d95117c46533141b9b/checks/34706443170/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
....................................................................i....

failures:

---- library/core/src/alloc/global.rs - alloc::global::GlobalAlloc (line 22) stdout ----
error: the `#[global_allocator]` in this crate conflicts with global allocator in: std
error: aborting due to 1 previous error

Couldn't compile the test.

```

</details>

bjorn3 commented:
@bors try

rust-log-analyzer commented:
The job **`x86_64-msvc`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/12431531590/job/34709155774) [(plain)](https://github.com/rust-lang-ci/rust/commit/3659c428d54e62b8a890cc8fb55f665040d0ed52/checks/34709155774/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Updating files:  98% (49896/50914)
Updating files:  99% (50405/50914)
Updating files: 100% (50914/50914)
Updating files: 100% (50914/50914), done.
branch 'try' set up to track 'origin/try'.
Switched to a new branch 'try'
[command]"C:\Program Files\Git\bin\git.exe" log -1 --format=%H
3659c428d54e62b8a890cc8fb55f665040d0ed52
##[group]Run src/ci/scripts/setup-environment.sh
src/ci/scripts/setup-environment.sh
---
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config remote.upstream.fetch=+refs/heads/*:refs/remotes/upstream/*
file:.git/config submodule.library/backtrace.active=true
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
---
[RUSTC-TIMING] alloc test:false 7.061
[RUSTC-TIMING] hashbrown test:false 1.571
error: linking with `link.exe` failed: exit code: 1120
  |
  = note: "C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.42.34433\\bin\\HostX64\\x64\\link.exe" "/DEF:C:\\a\\_temp\\msys64\\tmp\\rustcZPtXff\\lib.def" "/NOLOGO" "C:\\a\\_temp\\msys64\\tmp\\rustcZPtXff\\symbols.o" "<1 object files omitted>" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-std\\x86_64-pc-windows-msvc\\release\\deps\\std-841d3ce69c23e140.6jzzfk6c3meyins4e40nuv3tl.rcgu.rmeta" "<1 object files omitted>" "kernel32.lib" "kernel32.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-std\\x86_64-pc-windows-msvc\\release\\deps/{libpanic_unwind-352ed0c904cab217.rlib,libwindows_targets-6bfa9ea28a1d6264.rlib,librustc_demangle-6a9a2581301dee4a.rlib,libstd_detect-0d33bcfb9d85207c.rlib,libhashbrown-8afd51a1b1451245.rlib,librustc_std_workspace_alloc-a2e4584ee5a070de.rlib,libunwind-af589fe8b7fd9ddd.rlib,libcfg_if-be134a3cd4d399a9.rlib,liballoc-a65a7b1cf7120177.rlib,librustc_std_workspace_core-ae2fade04ec27265.rlib,libcore-d373f053f076a95d.rlib,libcompiler_builtins-b7ee4c0bb31856ae.rlib}" "advapi32.lib" "ntdll.lib" "userenv.lib" "ws2_32.lib" "dbghelp.lib" "/defaultlib:libcmt" "C:\\a\\_temp\\msys64\\tmp\\rustcZPtXff\\bcryptprimitives.dll_imports.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcZPtXff\\api-ms-win-core-synch-l1-2-0.dll_imports.lib" "/NXCOMPAT" "/LIBPATH:C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.42.34433\\atlmfc\\lib\\x64" "/LIBPATH:C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-std\\x86_64-pc-windows-msvc\\release\\build\\compiler_builtins-e8111bc73de9b80a\\out" "/OUT:C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-std\\x86_64-pc-windows-msvc\\release\\deps\\std-841d3ce69c23e140.dll" "/OPT:REF,ICF" "/DLL" "/IMPLIB:C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-std\\x86_64-pc-windows-msvc\\release\\deps\\std-841d3ce69c23e140.dll.lib" "/DEBUG" "/PDBALTPATH:%_PDB%"
  = note: lib.def : error LNK2001: unresolved external symbol __rust_alloc␍
          lib.def : error LNK2001: unresolved external symbol __rust_alloc_zeroed␍
          lib.def : error LNK2001: unresolved external symbol __rust_dealloc␍
          lib.def : error LNK2001: unresolved external symbol __rust_realloc␍
          lib.def : error LNK2001: unresolved external symbol __rust_realloc␍
          C:\a\rust\rust\build\x86_64-pc-windows-msvc\stage1-std\x86_64-pc-windows-msvc\release\deps\std-841d3ce69c23e140.dll.lib : fatal error LNK1120: 4 unresolved externals␍
          

[RUSTC-TIMING] std test:false 19.600
error: could not compile `std` (lib) due to 1 previous error
Build completed unsuccessfully in 0:19:06
make: *** [Makefile:106: ci-msvc-ps1] Error 1
  network time: Fri, 20 Dec 2024 13:03:25 GMT
##[error]Process completed with exit code 2.
Post job cleanup.
[command]"C:\Program Files\Git\bin\git.exe" version
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12431531590/job/34710313165)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/12431531590/job/34710313165","builder_name":"checks-actions"} -->

bjorn3 commented:
@bors try

bors commented:
:hourglass: Trying commit 3e25d5ab0a695448a6ffd9de411b7a1fe3ea7215 with merge f0b034562a44d58633af9414efbd4c6416626e83...
<!-- homu: {"type":"TryBuildStarted","head_sha":"3e25d5ab0a695448a6ffd9de411b7a1fe3ea7215","merge_sha":"f0b034562a44d58633af9414efbd4c6416626e83"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12431499593/job/34709061921) [(plain)](https://github.com/rust-lang/rust/commit/3e25d5ab0a695448a6ffd9de411b7a1fe3ea7215/checks/34709061921/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#22 exporting to docker image format
#22 sending tarball 27.0s done
#22 DONE 32.3s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.83s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
---- [run-make] tests/run-make/bin-emit-no-symbols stdout ----

error: rmake recipe failed to complete
status: exit status: 101
command: cd "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/bin-emit-no-symbols/rmake_out" && env -u RUSTFLAGS AR="ar" CARGO="/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools-bin/cargo" CC="cc" CC_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC -m64" CXX="c++" CXX_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC -m64" HOST_RPATH_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools-bin:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" LD_LIB_PATH_ENVVAR="LD_LIBRARY_PATH" LLVM_BIN_DIR="/usr/lib/llvm-18/bin" LLVM_COMPONENTS="aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgputargetmca amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen codegentypes core coroutines coverage debuginfobtf debuginfocodeview debuginfodwarf debuginfogsym debuginfologicalview debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker dwarflinkerclassic dwarflinkerparallel dwp engine executionengine extensions filecheck frontenddriver frontendhlsl frontendoffloading frontendopenacc frontendopenmp fuzzercli fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo hipstdpar instcombine instrumentation interfacestub interpreter ipo irprinter irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker loongarch loongarchasmparser loongarchcodegen loongarchdesc loongarchdisassembler loongarchinfo lto m68k m68kasmparser m68kcodegen m68kdesc m68kdisassembler m68kinfo mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts objcopy object objectyaml option orcdebugging orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvtargetmca runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target targetparser textapi textapibinaryreader transformutils ve veasmparser vecodegen vectorize vedesc vedisassembler veinfo webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo webassemblyutils windowsdriver windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86targetmca xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray xtensa xtensaasmparser xtensacodegen xtensadesc xtensadisassembler xtensainfo" LLVM_FILECHECK="/usr/lib/llvm-18/bin/FileCheck" NODE="/usr/bin/node" PYTHON="/usr/bin/python3" RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" RUSTDOC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" SOURCE_ROOT="/checkout" TARGET="x86_64-unknown-linux-gnu" TARGET_RPATH_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" TARGET_RPATH_ENV="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/bin-emit-no-symbols/rmake_out:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/bin-emit-no-symbols/rmake"
--- stderr -------------------------------
assert_contains:
assert_contains:
=== HAYSTACK ===

Symbol table '.symtab' contains 21 entries:
   Num:    Value          Size Type    Bind   Vis       Ndx Name
     0: 0000000000000000     0 NOTYPE  LOCAL  DEFAULT   UND 
     1: 0000000000000000     0 FILE    LOCAL  DEFAULT   ABS app.1035310be4a0b0e-cgu.0
     2: 0000000000000000     0 SECTION LOCAL  DEFAULT     3 .text._ZN4core5alloc6layout6Layout25from_size_align_unchecked17ha1e84972d16b4448E
     3: 0000000000000000    46 FUNC    LOCAL  DEFAULT     3 _ZN4core5alloc6layout6Layout25from_size_align_unchecked17ha1e84972d16b4448E
     4: 0000000000000000    58 FUNC    LOCAL  DEFAULT     5 _ZN4core5alloc6layout6Layout25from_size_align_unchecked18precondition_check17h9600c514a9dbf053E
     5: 0000000000000000     0 SECTION LOCAL  DEFAULT     5 .text._ZN4core5alloc6layout6Layout25from_size_align_unchecked18precondition_check17h9600c514a9dbf053E
     6: 0000000000000000     0 SECTION LOCAL  DEFAULT     7 .gcc_except_table._ZN4core5alloc6layout6Layout25from_size_align_unchecked18precondition_check17h9600c514a9dbf053E
     7: 0000000000000000     0 NOTYPE  LOCAL  DEFAULT     7 GCC_except_table1
     8: 0000000000000000     0 SECTION LOCAL  DEFAULT     8 .text.rust_begin_unwind
     9: 0000000000000000     0 SECTION LOCAL  DEFAULT     9 .text.rust_eh_personality
    10: 0000000000000000     0 SECTION LOCAL  DEFAULT    10 .text._ZN3app3oom17h8541f5ec76ff41b1E
    11: 0000000000000000     4 FUNC    LOCAL  DEFAULT    10 _ZN3app3oom17h8541f5ec76ff41b1E
    12: 0000000000000000     0 SECTION LOCAL  DEFAULT    11 .text.__rust_alloc_error_handler
    13: 0000000000000000     0 SECTION LOCAL  DEFAULT    13 .rodata..L__unnamed_1
    14: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT   UND _RNvMNtNtCsbXpABHcC73X_4core5alloc6layoutNtB2_6Layout19is_size_align_valid
    15: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT   UND _RNvNtCsbXpABHcC73X_4core9panicking19panic_cannot_unwind
    16: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT   UND _RNvNtCsbXpABHcC73X_4core9panicking14panic_nounwind
    17: 0000000000000000     4 FUNC    GLOBAL HIDDEN      8 rust_begin_unwind
    18: 0000000000000000     1 FUNC    GLOBAL HIDDEN      9 rust_eh_personality
    19: 0000000000000000    17 FUNC    GLOBAL HIDDEN     11 __rust_alloc_error_handler
    20: 0000000000000000     8 OBJECT  WEAK   HIDDEN     15 DW.ref.rust_eh_personality
=== NEEDLE ===
rust_begin_unwind
assert_contains:
assert_contains:
=== HAYSTACK ===

Symbol table '.symtab' contains 21 entries:
   Num:    Value          Size Type    Bind   Vis       Ndx Name
     0: 0000000000000000     0 NOTYPE  LOCAL  DEFAULT   UND 
     1: 0000000000000000     0 FILE    LOCAL  DEFAULT   ABS app.1035310be4a0b0e-cgu.0
     2: 0000000000000000     0 SECTION LOCAL  DEFAULT     3 .text._ZN4core5alloc6layout6Layout25from_size_align_unchecked17ha1e84972d16b4448E
     3: 0000000000000000    46 FUNC    LOCAL  DEFAULT     3 _ZN4core5alloc6layout6Layout25from_size_align_unchecked17ha1e84972d16b4448E
     4: 0000000000000000    58 FUNC    LOCAL  DEFAULT     5 _ZN4core5alloc6layout6Layout25from_size_align_unchecked18precondition_check17h9600c514a9dbf053E
     5: 0000000000000000     0 SECTION LOCAL  DEFAULT     5 .text._ZN4core5alloc6layout6Layout25from_size_align_unchecked18precondition_check17h9600c514a9dbf053E
     6: 0000000000000000     0 SECTION LOCAL  DEFAULT     7 .gcc_except_table._ZN4core5alloc6layout6Layout25from_size_align_unchecked18precondition_check17h9600c514a9dbf053E
     7: 0000000000000000     0 NOTYPE  LOCAL  DEFAULT     7 GCC_except_table1
     8: 0000000000000000     0 SECTION LOCAL  DEFAULT     8 .text.rust_begin_unwind
     9: 0000000000000000     0 SECTION LOCAL  DEFAULT     9 .text.rust_eh_personality
    10: 0000000000000000     0 SECTION LOCAL  DEFAULT    10 .text._ZN3app3oom17h8541f5ec76ff41b1E
    11: 0000000000000000     4 FUNC    LOCAL  DEFAULT    10 _ZN3app3oom17h8541f5ec76ff41b1E
    12: 0000000000000000     0 SECTION LOCAL  DEFAULT    11 .text.__rust_alloc_error_handler
    13: 0000000000000000     0 SECTION LOCAL  DEFAULT    13 .rodata..L__unnamed_1
    14: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT   UND _RNvMNtNtCsbXpABHcC73X_4core5alloc6layoutNtB2_6Layout19is_size_align_valid
    15: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT   UND _RNvNtCsbXpABHcC73X_4core9panicking19panic_cannot_unwind
    16: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT   UND _RNvNtCsbXpABHcC73X_4core9panicking14panic_nounwind
    17: 0000000000000000     4 FUNC    GLOBAL HIDDEN      8 rust_begin_unwind
    18: 0000000000000000     1 FUNC    GLOBAL HIDDEN      9 rust_eh_personality
    19: 0000000000000000    17 FUNC    GLOBAL HIDDEN     11 __rust_alloc_error_handler
    20: 0000000000000000     8 OBJECT  WEAK   HIDDEN     15 DW.ref.rust_eh_personality
=== NEEDLE ===
rust_eh_personality
assert_contains:
assert_contains:
=== HAYSTACK ===

Symbol table '.symtab' contains 21 entries:
   Num:    Value          Size Type    Bind   Vis       Ndx Name
     0: 0000000000000000     0 NOTYPE  LOCAL  DEFAULT   UND 
     1: 0000000000000000     0 FILE    LOCAL  DEFAULT   ABS app.1035310be4a0b0e-cgu.0
     2: 0000000000000000     0 SECTION LOCAL  DEFAULT     3 .text._ZN4core5alloc6layout6Layout25from_size_align_unchecked17ha1e84972d16b4448E
     3: 0000000000000000    46 FUNC    LOCAL  DEFAULT     3 _ZN4core5alloc6layout6Layout25from_size_align_unchecked17ha1e84972d16b4448E
     4: 0000000000000000    58 FUNC    LOCAL  DEFAULT     5 _ZN4core5alloc6layout6Layout25from_size_align_unchecked18precondition_check17h9600c514a9dbf053E
     5: 0000000000000000     0 SECTION LOCAL  DEFAULT     5 .text._ZN4core5alloc6layout6Layout25from_size_align_unchecked18precondition_check17h9600c514a9dbf053E
     6: 0000000000000000     0 SECTION LOCAL  DEFAULT     7 .gcc_except_table._ZN4core5alloc6layout6Layout25from_size_align_unchecked18precondition_check17h9600c514a9dbf053E
     7: 0000000000000000     0 NOTYPE  LOCAL  DEFAULT     7 GCC_except_table1
     8: 0000000000000000     0 SECTION LOCAL  DEFAULT     8 .text.rust_begin_unwind
     9: 0000000000000000     0 SECTION LOCAL  DEFAULT     9 .text.rust_eh_personality
    10: 0000000000000000     0 SECTION LOCAL  DEFAULT    10 .text._ZN3app3oom17h8541f5ec76ff41b1E
    11: 0000000000000000     4 FUNC    LOCAL  DEFAULT    10 _ZN3app3oom17h8541f5ec76ff41b1E
    12: 0000000000000000     0 SECTION LOCAL  DEFAULT    11 .text.__rust_alloc_error_handler
    13: 0000000000000000     0 SECTION LOCAL  DEFAULT    13 .rodata..L__unnamed_1
    14: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT   UND _RNvMNtNtCsbXpABHcC73X_4core5alloc6layoutNtB2_6Layout19is_size_align_valid
    15: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT   UND _RNvNtCsbXpABHcC73X_4core9panicking19panic_cannot_unwind
    16: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT   UND _RNvNtCsbXpABHcC73X_4core9panicking14panic_nounwind
    17: 0000000000000000     4 FUNC    GLOBAL HIDDEN      8 rust_begin_unwind
    18: 0000000000000000     1 FUNC    GLOBAL HIDDEN      9 rust_eh_personality
    19: 0000000000000000    17 FUNC    GLOBAL HIDDEN     11 __rust_alloc_error_handler
    20: 0000000000000000     8 OBJECT  WEAK   HIDDEN     15 DW.ref.rust_eh_personality
=== NEEDLE ===
__rg_oom
thread 'main' panicked at /checkout/tests/run-make/bin-emit-no-symbols/rmake.rs:15:9:
needle was not found in haystack
```

</details>

bjorn3 committed:
Update tests

rust-log-analyzer commented:
The job **`x86_64-apple-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/12432201626/job/34711125465) [(plain)](https://github.com/rust-lang-ci/rust/commit/f0b034562a44d58633af9414efbd4c6416626e83/checks/34711125465/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Updating files:  98% (50262/50914)
Updating files:  99% (50405/50914)
Updating files: 100% (50914/50914)
Updating files: 100% (50914/50914), done.
Switched to a new branch 'try'
branch 'try' set up to track 'origin/try'.
[command]/usr/local/bin/git log -1 --format=%H
f0b034562a44d58633af9414efbd4c6416626e83
##[group]Run src/ci/scripts/setup-environment.sh
src/ci/scripts/setup-environment.sh
---
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config remote.upstream.fetch=+refs/heads/*:refs/remotes/upstream/*
file:.git/config submodule.library/backtrace.active=true
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
---
---- [run-make] tests/run-make/bin-emit-no-symbols stdout ----

error: rmake recipe failed to complete
status: exit status: 101
command: cd "/Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/run-make/bin-emit-no-symbols/rmake_out" && env -u RUSTFLAGS AR="ar" CARGO="/Users/runner/work/rust/rust/build/x86_64-apple-darwin/stage1-tools-bin/cargo" CC="/Users/runner/work/rust/rust/clang+llvm-15.0.7-x86_64-apple-darwin21.0/bin/clang" CC_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC -m64 --target=x86_64-apple-macosx10.12" CXX="/Users/runner/work/rust/rust/clang+llvm-15.0.7-x86_64-apple-darwin21.0/bin/clang++" CXX_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC -m64 --target=x86_64-apple-macosx10.12 -stdlib=libc++" DYLD_LIBRARY_PATH="/Users/runner/work/rust/rust/build/x86_64-apple-darwin/stage0-bootstrap-tools/x86_64-apple-darwin/release/deps:/Users/runner/work/rust/rust/build/x86_64-apple-darwin/stage0/lib:/Users/runner/work/rust/rust/build/x86_64-apple-darwin/stage2-tools-bin:/Users/runner/work/rust/rust/build/x86_64-apple-darwin/stage2/lib/rustlib/x86_64-apple-darwin/lib" HOST_RPATH_DIR="/Users/runner/work/rust/rust/build/x86_64-apple-darwin/stage2/lib" LD_LIB_PATH_ENVVAR="DYLD_LIBRARY_PATH" LLVM_BIN_DIR="/Users/runner/work/rust/rust/build/x86_64-apple-darwin/ci-llvm/bin" LLVM_COMPONENTS="aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen codegendata codegentypes core coroutines coverage csky cskyasmparser cskycodegen cskydesc cskydisassembler cskyinfo debuginfobtf debuginfocodeview debuginfodwarf debuginfogsym debuginfologicalview debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker dwarflinkerclassic dwarflinkerparallel dwp engine executionengine extensions filecheck frontenddriver frontendhlsl frontendoffloading frontendopenacc frontendopenmp fuzzercli fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo hipstdpar instcombine instrumentation interfacestub interpreter ipo irprinter irreader jitlink libdriver lineeditor linker loongarch loongarchasmparser loongarchcodegen loongarchdesc loongarchdisassembler loongarchinfo lto m68k m68kasmparser m68kcodegen m68kdesc m68kdisassembler m68kinfo mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts objcopy object objectyaml option orcdebugging orcjit orcshared orctargetprocess passes powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvtargetmca runtimedyld sandboxir scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target targetparser textapi textapibinaryreader transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo webassemblyutils windowsdriver windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86targetmca xray xtensa xtensaasmparser xtensacodegen xtensadesc xtensadisassembler xtensainfo" LLVM_FILECHECK="/Users/runner/work/rust/rust/build/x86_64-apple-darwin/ci-llvm/bin/FileCheck" NODE="/usr/local/bin/node" PYTHON="/usr/bin/python3" RUSTC="/Users/runner/work/rust/rust/build/x86_64-apple-darwin/stage2/bin/rustc" RUSTDOC="/Users/runner/work/rust/rust/build/x86_64-apple-darwin/stage2/bin/rustdoc" SOURCE_ROOT="/Users/runner/work/rust/rust" TARGET="x86_64-apple-darwin" TARGET_RPATH_DIR="/Users/runner/work/rust/rust/build/x86_64-apple-darwin/stage2/lib/rustlib/x86_64-apple-darwin/lib" TARGET_RPATH_ENV="/Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/run-make/bin-emit-no-symbols/rmake_out:/Users/runner/work/rust/rust/build/x86_64-apple-darwin/stage0-bootstrap-tools/x86_64-apple-darwin/release/deps:/Users/runner/work/rust/rust/build/x86_64-apple-darwin/stage0/lib" "/Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/run-make/bin-emit-no-symbols/rmake"
--- stderr -------------------------------
assert_contains:
assert_contains:
=== HAYSTACK ===
File: app.o
File: app.o
Format: Mach-O 64-bit x86-64
Arch: x86_64
AddressSize: 64bit
Symbols [
  Symbol {
    Name: __ZN4core5alloc6layout6Layout25from_size_align_unchecked17h0a520882f3b2cc3dE (69)
    Type: Section (0xE)
    Section: __text (0x1)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
    Value: 0x0
  }
  Symbol {
  Symbol {
    Name: __ZN4core5alloc6layout6Layout25from_size_align_unchecked18precondition_check17h16e2b060a144ff12E (367)
    Type: Section (0xE)
    Section: __text (0x1)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
    Value: 0x30
  }
  Symbol {
  Symbol {
    Name: l___unnamed_1 (482)
    Type: Section (0xE)
    Section: __const (0x3)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
    Value: 0xC4
  }
  Symbol {
    Name: GCC_except_table1 (464)
    Name: GCC_except_table1 (464)
    Type: Section (0xE)
    Section: __gcc_except_tab (0x2)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
    Value: 0xB4
  }
  Symbol {
  Symbol {
    Name: __ZN3app3oom17h8cfc2a66c883d686E (278)
    Type: Section (0xE)
    Section: __text (0x1)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
    Value: 0x90
  }
  Symbol {
    Name: ___rust_alloc_error_handler (22)
    Name: ___rust_alloc_error_handler (22)
    PrivateExtern
    Extern
    Type: Section (0xE)
    Section: __text (0x1)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
    Value: 0xA0
  }
  Symbol {
    Name: _rust_begin_unwind (50)
    Name: _rust_begin_unwind (50)
    PrivateExtern
    Extern
    Type: Section (0xE)
    Section: __text (0x1)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
    Value: 0x70
  }
  Symbol {
  Symbol {
    Name: _rust_eh_personality (1)
    PrivateExtern
    Type: Section (0xE)
    Type: Section (0xE)
    Section: __text (0x1)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
    Value: 0x80
  }
  Symbol {
  Symbol {
    Name: __ZN4core5alloc6layout6Layout19is_size_align_valid17h56e3ffe8dcbf744cE (146)
    Type: Undef (0x0)
    Section:  (0x0)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
    Flags [ (0x0)
    ]
    Value: 0x0
  }
  Symbol {
    Name: __ZN4core9panicking14panic_nounwind17h6c494ce30a2dc7f2E (311)
    Type: Undef (0x0)
    Section:  (0x0)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
    Flags [ (0x0)
    ]
    Value: 0x0
  }
  Symbol {
    Name: __ZN4core9panicking19panic_cannot_unwind17haff82019bed171a8E (217)
    Type: Undef (0x0)
    Section:  (0x0)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
---

=== NEEDLE ===
rust_begin_unwind
assert_contains:
=== HAYSTACK ===
File: app.o
File: app.o
Format: Mach-O 64-bit x86-64
Arch: x86_64
AddressSize: 64bit
Symbols [
  Symbol {
    Name: __ZN4core5alloc6layout6Layout25from_size_align_unchecked17h0a520882f3b2cc3dE (69)
    Type: Section (0xE)
    Section: __text (0x1)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
    Value: 0x0
  }
  Symbol {
  Symbol {
    Name: __ZN4core5alloc6layout6Layout25from_size_align_unchecked18precondition_check17h16e2b060a144ff12E (367)
    Type: Section (0xE)
    Section: __text (0x1)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
    Value: 0x30
  }
  Symbol {
  Symbol {
    Name: l___unnamed_1 (482)
    Type: Section (0xE)
    Section: __const (0x3)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
    Value: 0xC4
  }
  Symbol {
    Name: GCC_except_table1 (464)
    Name: GCC_except_table1 (464)
    Type: Section (0xE)
    Section: __gcc_except_tab (0x2)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
    Value: 0xB4
  }
  Symbol {
  Symbol {
    Name: __ZN3app3oom17h8cfc2a66c883d686E (278)
    Type: Section (0xE)
    Section: __text (0x1)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
    Value: 0x90
  }
  Symbol {
    Name: ___rust_alloc_error_handler (22)
    Name: ___rust_alloc_error_handler (22)
    PrivateExtern
    Extern
    Type: Section (0xE)
    Section: __text (0x1)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
    Value: 0xA0
  }
  Symbol {
    Name: _rust_begin_unwind (50)
    Name: _rust_begin_unwind (50)
    PrivateExtern
    Extern
    Type: Section (0xE)
    Section: __text (0x1)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
    Value: 0x70
  }
  Symbol {
  Symbol {
    Name: _rust_eh_personality (1)
    PrivateExtern
    Type: Section (0xE)
    Type: Section (0xE)
    Section: __text (0x1)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
    Value: 0x80
  }
  Symbol {
  Symbol {
    Name: __ZN4core5alloc6layout6Layout19is_size_align_valid17h56e3ffe8dcbf744cE (146)
    Type: Undef (0x0)
    Section:  (0x0)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
    Flags [ (0x0)
    ]
    Value: 0x0
  }
  Symbol {
    Name: __ZN4core9panicking14panic_nounwind17h6c494ce30a2dc7f2E (311)
    Type: Undef (0x0)
    Section:  (0x0)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
    Flags [ (0x0)
    ]
    Value: 0x0
  }
  Symbol {
    Name: __ZN4core9panicking19panic_cannot_unwind17haff82019bed171a8E (217)
    Type: Undef (0x0)
    Section:  (0x0)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
---

=== NEEDLE ===
rust_eh_personality
assert_contains:
=== HAYSTACK ===
File: app.o
File: app.o
Format: Mach-O 64-bit x86-64
Arch: x86_64
AddressSize: 64bit
Symbols [
  Symbol {
    Name: __ZN4core5alloc6layout6Layout25from_size_align_unchecked17h0a520882f3b2cc3dE (69)
    Type: Section (0xE)
    Section: __text (0x1)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
    Value: 0x0
  }
  Symbol {
  Symbol {
    Name: __ZN4core5alloc6layout6Layout25from_size_align_unchecked18precondition_check17h16e2b060a144ff12E (367)
    Type: Section (0xE)
    Section: __text (0x1)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
    Value: 0x30
  }
  Symbol {
  Symbol {
    Name: l___unnamed_1 (482)
    Type: Section (0xE)
    Section: __const (0x3)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
    Value: 0xC4
  }
  Symbol {
    Name: GCC_except_table1 (464)
    Name: GCC_except_table1 (464)
    Type: Section (0xE)
    Section: __gcc_except_tab (0x2)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
    Value: 0xB4
  }
  Symbol {
  Symbol {
    Name: __ZN3app3oom17h8cfc2a66c883d686E (278)
    Type: Section (0xE)
    Section: __text (0x1)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
    Value: 0x90
  }
  Symbol {
    Name: ___rust_alloc_error_handler (22)
    Name: ___rust_alloc_error_handler (22)
    PrivateExtern
    Extern
    Type: Section (0xE)
    Section: __text (0x1)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
    Value: 0xA0
  }
  Symbol {
    Name: _rust_begin_unwind (50)
    Name: _rust_begin_unwind (50)
    PrivateExtern
    Extern
    Type: Section (0xE)
    Section: __text (0x1)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
    Value: 0x70
  }
  Symbol {
  Symbol {
    Name: _rust_eh_personality (1)
    PrivateExtern
    Type: Section (0xE)
    Type: Section (0xE)
    Section: __text (0x1)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
    Value: 0x80
  }
  Symbol {
  Symbol {
    Name: __ZN4core5alloc6layout6Layout19is_size_align_valid17h56e3ffe8dcbf744cE (146)
    Type: Undef (0x0)
    Section:  (0x0)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
    Flags [ (0x0)
    ]
    Value: 0x0
  }
  Symbol {
    Name: __ZN4core9panicking14panic_nounwind17h6c494ce30a2dc7f2E (311)
    Type: Undef (0x0)
    Section:  (0x0)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
    Flags [ (0x0)
    ]
    Value: 0x0
  }
  Symbol {
    Name: __ZN4core9panicking19panic_cannot_unwind17haff82019bed171a8E (217)
    Type: Undef (0x0)
    Section:  (0x0)
    RefType: UndefinedNonLazy (0x0)
    Flags [ (0x0)
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #134604) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bjorn3 commented:
@bors try

bors commented:
:hourglass: Trying commit 7952352980f2d5326cdb80ccbd88d4fedbf430f3 with merge d5915d9567206920e25d38421ce09513599a9d53...
<!-- homu: {"type":"TryBuildStarted","head_sha":"7952352980f2d5326cdb80ccbd88d4fedbf430f3","merge_sha":"d5915d9567206920e25d38421ce09513599a9d53"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12830536756/job/35789094781)
Build commit: d5915d9567206920e25d38421ce09513599a9d53 (`d5915d9567206920e25d38421ce09513599a9d53`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/12830536756/job/35789094781"},"merge_sha":"d5915d9567206920e25d38421ce09513599a9d53"} -->

bjorn3 commented:
And for windows.

@bors try

bors commented:
:hourglass: Trying commit 7952352980f2d5326cdb80ccbd88d4fedbf430f3 with merge 0b444637461eed9bab0f10716f94c3b4b969d28d...
<!-- homu: {"type":"TryBuildStarted","head_sha":"7952352980f2d5326cdb80ccbd88d4fedbf430f3","merge_sha":"0b444637461eed9bab0f10716f94c3b4b969d28d"} -->

rust-log-analyzer commented:
The job **`x86_64-msvc`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/12836757750/job/35799081809) [(plain)](https://github.com/rust-lang-ci/rust/commit/0b444637461eed9bab0f10716f94c3b4b969d28d/checks/35799081809/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Updating files:  99% (51022/51537)
Updating files:  99% (51382/51537)
Updating files: 100% (51537/51537)
Updating files: 100% (51537/51537), done.
branch 'try' set up to track 'origin/try'.
Switched to a new branch 'try'
[command]"C:\Program Files\Git\bin\git.exe" log -1 --format=%H
0b444637461eed9bab0f10716f94c3b4b969d28d
##[group]Run src/ci/scripts/setup-environment.sh
src/ci/scripts/setup-environment.sh
---
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config remote.upstream.fetch=+refs/heads/*:refs/remotes/upstream/*
file:.git/config submodule.library/backtrace.active=true
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
---
[RUSTC-TIMING] alloc test:false 7.039
[RUSTC-TIMING] hashbrown test:false 1.687
error: linking with `link.exe` failed: exit code: 1120
  |
  = note: "C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.42.34433\\bin\\HostX64\\x64\\link.exe" "/DEF:C:\\a\\_temp\\msys64\\tmp\\rustcIYnPuy\\lib.def" "/NOLOGO" "C:\\a\\_temp\\msys64\\tmp\\rustcIYnPuy\\symbols.o" "<1 object files omitted>" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-std\\x86_64-pc-windows-msvc\\release\\deps\\std-7194203398611eb3.aog2zyr6dovrqwqus3ht0et76.rcgu.rmeta" "<1 object files omitted>" "kernel32.lib" "kernel32.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-std\\x86_64-pc-windows-msvc\\release\\deps/{libpanic_unwind-3fcd4d4f4c698c86.rlib,libwindows_targets-1cb2c690ed000745.rlib,librustc_demangle-71a11a2fd5cfde68.rlib,libstd_detect-30988aadaac3204a.rlib,libhashbrown-e7912727a443725d.rlib,librustc_std_workspace_alloc-3154bf3e0fa00acc.rlib,libunwind-e4f6a8f25c074a7f.rlib,libcfg_if-351ec6b29335afed.rlib,liballoc-3864fe2f128a16c1.rlib,librustc_std_workspace_core-e7a676e276d8d181.rlib,libcore-4125786f7157f39c.rlib,libcompiler_builtins-f4d65ae9a6856a64.rlib}" "advapi32.lib" "ntdll.lib" "userenv.lib" "ws2_32.lib" "dbghelp.lib" "/defaultlib:libcmt" "C:\\a\\_temp\\msys64\\tmp\\rustcIYnPuy\\bcryptprimitives.dll_imports.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcIYnPuy\\api-ms-win-core-synch-l1-2-0.dll_imports.lib" "/NXCOMPAT" "/LIBPATH:C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.42.34433\\atlmfc\\lib\\x64" "/LIBPATH:C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-std\\x86_64-pc-windows-msvc\\release\\build\\compiler_builtins-9305a008fbe857ee\\out" "/OUT:C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-std\\x86_64-pc-windows-msvc\\release\\deps\\std-7194203398611eb3.dll" "/OPT:REF,ICF" "/DLL" "/IMPLIB:C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-std\\x86_64-pc-windows-msvc\\release\\deps\\std-7194203398611eb3.dll.lib" "/DEBUG" "/PDBALTPATH:%_PDB%"
  = note: lib.def : error LNK2001: unresolved external symbol __rust_alloc␍
          lib.def : error LNK2001: unresolved external symbol __rust_alloc_zeroed␍
          lib.def : error LNK2001: unresolved external symbol __rust_dealloc␍
          lib.def : error LNK2001: unresolved external symbol __rust_realloc␍
          lib.def : error LNK2001: unresolved external symbol __rust_realloc␍
          C:\a\rust\rust\build\x86_64-pc-windows-msvc\stage1-std\x86_64-pc-windows-msvc\release\deps\std-7194203398611eb3.dll.lib : fatal error LNK1120: 4 unresolved externals␍
          

[RUSTC-TIMING] std test:false 20.571
error: could not compile `std` (lib) due to 1 previous error
Build completed unsuccessfully in 0:18:43
make: *** [Makefile:106: ci-msvc-ps1] Error 1
  network time: Fri, 17 Jan 2025 21:30:33 GMT
##[error]Process completed with exit code 2.
Post job cleanup.
[command]"C:\Program Files\Git\bin\git.exe" version
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12836757750/job/35800286964)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/12836757750/job/35800286964","builder_name":"checks-actions"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #136035) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

ChrisDenton commented:
I don't think anyone has worked on Rust's support for Windows weak linkage yet? Maybe `extern_weak` would work on Windows but I'm not even sure that's hooked up properly yet.

IIRC `#[linkage]` is just forwarding straight to LLVM and not always correctly.

bors commented:
:umbrella: The latest upstream changes (presumably #136751) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`mingw-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15678193674/job/44163388172) [(plain)](https://github.com/rust-lang/rust/commit/dcf99619c2ffa80a0926c31cc2a818baf34e2938/checks/44163388172/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

error[E0308]: mismatched types
   --> src/tools/miri/src/shims/foreign_items.rs:54:69
    |
54  |         match this.emulate_foreign_item_inner(link_name, abi, args, &dest)? {
    |                    --------------------------                       ^^^^^ expected `&MPlaceTy<'_, Provenance>`, found `&&PlaceTy<'_, Provenance>`
    |                    |
    |                    arguments to this method are incorrect
    |
    = note: expected reference `&rustc_const_eval::interpret::MPlaceTy<'_, machine::Provenance>`
               found reference `&&rustc_const_eval::interpret::PlaceTy<'tcx, machine::Provenance>`
---
    |
130 |                     if symbol_name == link_name {
    |                                    ^^ no implementation for `&str == rustc_span::Symbol`
    |
    = help: the trait `std::cmp::PartialEq<rustc_span::Symbol>` is not implemented for `&str`
    = help: the following other types implement trait `std::cmp::PartialEq<Rhs>`:
              `&str` implements `std::cmp::PartialEq<ByteString>`
              `&str` implements `std::cmp::PartialEq<Cow<'_, str>>`
              `&str` implements `std::cmp::PartialEq<OsString>`
              `&str` implements `std::cmp::PartialEq<bstr::bstr::BStr>`
              `&str` implements `std::cmp::PartialEq<bstr::bstring::BString>`
              `&str` implements `std::cmp::PartialEq<rustc_target::json::Json>`
              `&str` implements `std::cmp::PartialEq<std::bstr::ByteStr>`
              `&str` implements `std::cmp::PartialEq<std::string::String>`
            and 13 others

error[E0433]: failed to resolve: use of undeclared type `Conv`
   --> src/tools/miri/src/shims/foreign_items.rs:509:61
    |
509 |                     let [size, align] = ecx.check_shim(abi, Conv::Rust, link_name, args)?;
    |                                                             ^^^^ use of undeclared type `Conv`

error[E0433]: failed to resolve: use of undeclared type `Conv`
   --> src/tools/miri/src/shims/foreign_items.rs:544:62
    |
544 |                     let [size, align] = this.check_shim(abi, Conv::Rust, link_name, args)?;
    |                                                              ^^^^ use of undeclared type `Conv`

error[E0433]: failed to resolve: use of undeclared type `Conv`
   --> src/tools/miri/src/shims/foreign_items.rs:571:45
    |
571 |                         ecx.check_shim(abi, Conv::Rust, link_name, args)?;
    |                                             ^^^^ use of undeclared type `Conv`

error[E0433]: failed to resolve: use of undeclared type `Conv`
   --> src/tools/miri/src/shims/foreign_items.rs:606:46
    |
606 |                         this.check_shim(abi, Conv::Rust, link_name, args)?;
    |                                              ^^^^ use of undeclared type `Conv`

error[E0277]: can't compare `&str` with `rustc_span::Symbol`
   --> src/tools/miri/src/shims/foreign_items.rs:130:36
    |
130 |                     if symbol_name == link_name {
    |                                    ^^ no implementation for `&str == rustc_span::Symbol`
    |
    = help: the trait `std::cmp::PartialEq<rustc_span::Symbol>` is not implemented for `&str`
    = help: the following other types implement trait `std::cmp::PartialEq<Rhs>`:
              `&str` implements `std::cmp::PartialEq<ByteString>`
              `&str` implements `std::cmp::PartialEq<Cow<'_, str>>`
              `&str` implements `std::cmp::PartialEq<OsString>`
              `&str` implements `std::cmp::PartialEq<bstr::BStr>`
              `&str` implements `std::cmp::PartialEq<bstr::BString>`
              `&str` implements `std::cmp::PartialEq<bstr::bstr::BStr>`
              `&str` implements `std::cmp::PartialEq<bstr::bstring::BString>`
              `&str` implements `std::cmp::PartialEq<camino::Utf8Path>`
            and 24 others

Some errors have detailed explanations: E0277, E0308, E0433.
For more information about an error, try `rustc --explain E0277`.
[RUSTC-TIMING] miri test:false 3.888
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #141061) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bjorn3 committed:
Use a const for all __rust_alloc_error_handler name references

bjorn3 committed:
Support #[alloc_error_handler] without the allocator shim

Currently it is possible to avoid linking the allocator shim when
__rust_no_alloc_shim_is_unstable_v2 is defined when linking rlibs
directly as some build systems need. However this currently requires
liballoc to be compiled with --cfg no_global_oom_handling, which places
huge restrictions on what functions you can call and makes it impossible
to use libstd. Or alternatively you have to define
__rust_alloc_error_handler and (when using libstd)
__rust_alloc_error_handler_should_panic
using #[rustc_std_internal_symbol]. With this commit the former can be
replaced with #[alloc_error_handler].

Eventually the alloc_error_handler may either be removed entirely
(though the PR for that has been stale for years now) or we may start
using weak symbols for it instead. For the latter case this commit is a
prerequisite anyway.

rust-log-analyzer commented:
The job **`mingw-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15904848981/job/44856972459) [(plain)](https://github.com/rust-lang/rust/commit/63bb2bbbaa2d7e5ac3cd127268b09f6eb03a67b8/checks/44856972459/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

error[E0308]: mismatched types
   --> src/tools/miri/src/shims/foreign_items.rs:54:69
    |
54  |         match this.emulate_foreign_item_inner(link_name, abi, args, &dest)? {
    |                    --------------------------                       ^^^^^ expected `&MPlaceTy<'_, Provenance>`, found `&&PlaceTy<'_, Provenance>`
    |                    |
    |                    arguments to this method are incorrect
    |
    = note: expected reference `&rustc_const_eval::interpret::MPlaceTy<'_, machine::Provenance>`
               found reference `&&rustc_const_eval::interpret::PlaceTy<'tcx, machine::Provenance>`
---
    |
130 |                     if symbol_name == link_name {
    |                                    ^^ no implementation for `&str == rustc_span::Symbol`
    |
    = help: the trait `std::cmp::PartialEq<rustc_span::Symbol>` is not implemented for `&str`
    = help: the following other types implement trait `std::cmp::PartialEq<Rhs>`:
              `&str` implements `std::cmp::PartialEq<ByteString>`
              `&str` implements `std::cmp::PartialEq<Cow<'_, str>>`
              `&str` implements `std::cmp::PartialEq<OsString>`
              `&str` implements `std::cmp::PartialEq<bstr::bstr::BStr>`
              `&str` implements `std::cmp::PartialEq<bstr::bstring::BString>`
              `&str` implements `std::cmp::PartialEq<rustc_target::json::Json>`
              `&str` implements `std::cmp::PartialEq<std::bstr::ByteStr>`
              `&str` implements `std::cmp::PartialEq<std::string::String>`
            and 13 others

error[E0433]: failed to resolve: use of undeclared type `Conv`
   --> src/tools/miri/src/shims/foreign_items.rs:509:61
    |
509 |                     let [size, align] = ecx.check_shim(abi, Conv::Rust, link_name, args)?;
    |                                                             ^^^^ use of undeclared type `Conv`

error[E0433]: failed to resolve: use of undeclared type `Conv`
   --> src/tools/miri/src/shims/foreign_items.rs:544:62
    |
544 |                     let [size, align] = this.check_shim(abi, Conv::Rust, link_name, args)?;
    |                                                              ^^^^ use of undeclared type `Conv`

error[E0433]: failed to resolve: use of undeclared type `Conv`
   --> src/tools/miri/src/shims/foreign_items.rs:571:45
    |
571 |                         ecx.check_shim(abi, Conv::Rust, link_name, args)?;
    |                                             ^^^^ use of undeclared type `Conv`

error[E0433]: failed to resolve: use of undeclared type `Conv`
   --> src/tools/miri/src/shims/foreign_items.rs:606:46
    |
606 |                         this.check_shim(abi, Conv::Rust, link_name, args)?;
    |                                              ^^^^ use of undeclared type `Conv`

error[E0277]: can't compare `&str` with `rustc_span::Symbol`
   --> src/tools/miri/src/shims/foreign_items.rs:130:36
    |
130 |                     if symbol_name == link_name {
    |                                    ^^ no implementation for `&str == rustc_span::Symbol`
    |
    = help: the trait `std::cmp::PartialEq<rustc_span::Symbol>` is not implemented for `&str`
    = help: the following other types implement trait `std::cmp::PartialEq<Rhs>`:
              `&str` implements `std::cmp::PartialEq<ByteString>`
              `&str` implements `std::cmp::PartialEq<Cow<'_, str>>`
              `&str` implements `std::cmp::PartialEq<OsString>`
              `&str` implements `std::cmp::PartialEq<bstr::BStr>`
              `&str` implements `std::cmp::PartialEq<bstr::BString>`
              `&str` implements `std::cmp::PartialEq<bstr::bstr::BStr>`
              `&str` implements `std::cmp::PartialEq<bstr::bstring::BString>`
              `&str` implements `std::cmp::PartialEq<camino::Utf8Path>`
            and 24 others

Some errors have detailed explanations: E0277, E0308, E0433.
For more information about an error, try `rustc --explain E0277`.
[RUSTC-TIMING] miri test:false 3.641
```

</details>

rust-log-analyzer commented:
The job **`mingw-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15905424677/job/44858993851) [(plain)](https://github.com/rust-lang/rust/commit/1fe27c582e916b21a7a60772d367658d2b27d863/checks/44858993851/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

error[E0308]: mismatched types
   --> src/tools/miri/src/shims/foreign_items.rs:54:69
    |
54  |         match this.emulate_foreign_item_inner(link_name, abi, args, &dest)? {
    |                    --------------------------                       ^^^^^ expected `&MPlaceTy<'_, Provenance>`, found `&&PlaceTy<'_, Provenance>`
    |                    |
    |                    arguments to this method are incorrect
    |
    = note: expected reference `&rustc_const_eval::interpret::MPlaceTy<'_, machine::Provenance>`
               found reference `&&rustc_const_eval::interpret::PlaceTy<'tcx, machine::Provenance>`
---
    |
130 |                     if symbol_name == link_name {
    |                                    ^^ no implementation for `&str == rustc_span::Symbol`
    |
    = help: the trait `std::cmp::PartialEq<rustc_span::Symbol>` is not implemented for `&str`
    = help: the following other types implement trait `std::cmp::PartialEq<Rhs>`:
              `&str` implements `std::cmp::PartialEq<ByteString>`
              `&str` implements `std::cmp::PartialEq<Cow<'_, str>>`
              `&str` implements `std::cmp::PartialEq<OsString>`
              `&str` implements `std::cmp::PartialEq<bstr::bstr::BStr>`
              `&str` implements `std::cmp::PartialEq<bstr::bstring::BString>`
              `&str` implements `std::cmp::PartialEq<rustc_target::json::Json>`
              `&str` implements `std::cmp::PartialEq<std::bstr::ByteStr>`
              `&str` implements `std::cmp::PartialEq<std::string::String>`
            and 13 others

error[E0433]: failed to resolve: use of undeclared type `Conv`
   --> src/tools/miri/src/shims/foreign_items.rs:509:61
    |
509 |                     let [size, align] = ecx.check_shim(abi, Conv::Rust, link_name, args)?;
    |                                                             ^^^^ use of undeclared type `Conv`

error[E0433]: failed to resolve: use of undeclared type `Conv`
   --> src/tools/miri/src/shims/foreign_items.rs:544:62
    |
544 |                     let [size, align] = this.check_shim(abi, Conv::Rust, link_name, args)?;
    |                                                              ^^^^ use of undeclared type `Conv`

error[E0433]: failed to resolve: use of undeclared type `Conv`
   --> src/tools/miri/src/shims/foreign_items.rs:571:45
    |
571 |                         ecx.check_shim(abi, Conv::Rust, link_name, args)?;
    |                                             ^^^^ use of undeclared type `Conv`

error[E0433]: failed to resolve: use of undeclared type `Conv`
   --> src/tools/miri/src/shims/foreign_items.rs:606:46
    |
606 |                         this.check_shim(abi, Conv::Rust, link_name, args)?;
    |                                              ^^^^ use of undeclared type `Conv`

error[E0277]: can't compare `&str` with `rustc_span::Symbol`
   --> src/tools/miri/src/shims/foreign_items.rs:130:36
    |
130 |                     if symbol_name == link_name {
    |                                    ^^ no implementation for `&str == rustc_span::Symbol`
    |
    = help: the trait `std::cmp::PartialEq<rustc_span::Symbol>` is not implemented for `&str`
    = help: the following other types implement trait `std::cmp::PartialEq<Rhs>`:
              `&str` implements `std::cmp::PartialEq<ByteString>`
              `&str` implements `std::cmp::PartialEq<Cow<'_, str>>`
              `&str` implements `std::cmp::PartialEq<OsString>`
              `&str` implements `std::cmp::PartialEq<bstr::BStr>`
              `&str` implements `std::cmp::PartialEq<bstr::BString>`
              `&str` implements `std::cmp::PartialEq<bstr::bstr::BStr>`
              `&str` implements `std::cmp::PartialEq<bstr::bstring::BString>`
              `&str` implements `std::cmp::PartialEq<camino::Utf8Path>`
            and 24 others

Some errors have detailed explanations: E0277, E0308, E0433.
For more information about an error, try `rustc --explain E0277`.
[RUSTC-TIMING] miri test:false 3.757
```

</details>

bjorn3 commented:
Closing in favor of the EII work.
